### ISSUE TITLE
Implementing ML Emulator of Regression Model

### ISSUE THREAD

**Issue by BenjaminTJohnson** [JCSDA/CRTM_coef#11]
# PyTorch ML Emulator Implementation Plan

## Background & Motivation
To enable rapid prototyping of instrument coefficients, we need a modern, scalable, and maintainable PyTorch-based pipeline.

This approach raises concerns regarding deployment targets (ONNX vs. Fortran), the preservation of physical inductive biases (channel-specific weighting), and the migration path. This plan addresses those concerns directly.

## Scope & Impact
*   **Parallel Coexistence**: The new code will reside in a new directory (`ml_emulator/`) alongside the existing `ml_coef_gen/`. The legacy code will remain untouched until the new pipeline is fully validated.
*   **Data Source**: The pipeline will ingest the high-fidelity `TauProfile.nc` files generated by the `emulator/` Phase 1 process, ensuring a 1:1 comparison with traditional ODPS regression.

### 1. Data Ingestion
*   Develop a PyTorch `Dataset` (`dataset.py`) capable of parsing generalized `TauProfile.nc` files.
*   Extract features (Temperature, Pressure, Absorbers) and targets (Layer Optical Depths).

### 2. Model Architecture & Inductive Bias
*   **Generalized Dimensions**: The MLP's input/output dimensions will be dynamically sized based on the sensor configuration.
*   **Preserving Physics (Addressing Claude's Concern)**: We will not lose the inductive bias. Instead of hardcoding (e.g., `FOCUS_WET_CHANNELS = (19, 20, 21, 22)`), we will implement a **configuration-driven** architecture.
    *   **Component Separation**: The model will support predicting Dry and Wet components separately if defined in the config.
    *   **Weighted Loss**: The training loop will support a custom loss function where specific channels (or regimes) can be assigned higher penalty weights via the YAML configuration, replicating the behavior of `_focused_wet_sample_weight` without hardcoding sensor IDs.

### 3. Training Pipeline
*   Implement a standard PyTorch training loop (`train.py`) with GPU support (CUDA/MPS), validation splits, and learning rate scheduling (e.g., `ReduceLROnPlateau`).

### 4. Export & Deployment (The Migration Bridge)
To address the CRTM Fortran integration reality, we will support a dual-export strategy:
*   **Modern Target (ONNX)**: Export the trained PyTorch model to `.onnx`. This is the long-term goal for CRTM deployment via `onnxruntime` C-API bindings.
*   **Legacy Bridge (NetCDF)**: Implement an export utility (`export_legacy.py`) that extracts the weights, biases, and scaling factors from the PyTorch model and packs them into the exact same custom NetCDF format (`sensor_emulator.nc`) expected by the current Fortran implementation. This ensures zero downtime or broken dependencies for current CRTM consumers.

## Implementation Steps
1.  **Setup**: Create `ml_emulator/` and define the YAML configuration schema.
2.  **Data**: Implement and test the PyTorch `TauProfileDataset`.
3.  **Model**: Implement the configurable MLP and custom weighted loss function.
4.  **Train**: Build the training loop and verify convergence on a small subset of the ATMS data.
5.  **Export**: Implement both ONNX and Legacy NetCDF export functions.

## Verification
*   Train the new emulator on the ATMS N20 dataset.
*   Compare the resulting RMSE/MAE metrics against the metrics produced by the legacy `ml_coef_gen/` pipeline.
*   Verify that the generated legacy NetCDF file matches the structural expectations of the existing CRTM Fortran reader.

**Comment by BenjaminTJohnson** [comment]
Title: Implementation of PyTorch-based ResNet Emulator with Dynamic Channel Weighting

  We have successfully implemented a modern machine learning replacement for the traditional CRTM ODPS emulator within the ml_emulator/ directory. This pipeline is designed to handle high-channel-count   instruments like IASI (8,461 channels) with significantly higher accuracy and stability than traditional regression.

**Key Features:**
   * Architecture: Utilizes a deep ResNetEmulator (implemented in PyTorch 2.6) designed to learn subtle delta-transmittances. The residual connections allow the model to capture fine-scale spectral features that are often lost in traditional linear/polynomial regressions.
   * Dynamic Channel Weighting: Implemented a custom DynamicWeightedMSELoss function. This automatically identifies "hard" spectral channels during training and scales their loss penalty dynamically. This solved the "Worst Max Error" issue for IASI without requiring manual, sensor-specific weighting configurations.
   * Data Pipeline: A generalized TauProfileDataset that ingests standard `TauProfile.nc` files, ensuring a 1:1 ground-truth comparison with the traditional LBLRTM-based coefficient generation process.
   * Dual-Export Strategy:
       * ONNX Target: Direct export to .onnx for high-performance deployment via onnxruntime.
       * Legacy Bridge: An `export_legacy.py` utility that maintains backward compatibility with current Fortran NetCDF readers (with built-in architecture safety checks).
   * Fortran Integration: Developed a C-Bridge (`crtm_onnx_bridge.c`) and a Fortran test suite (`test_onnx_fortran.f90`) demonstrating how to call the exported ONNX models directly from the CRTM core using `iso_c_binding`.

  **Validation Results (IASI MetOp-A):**
   * Numerical Accuracy: Achieved a Global Avg RMSE of 0.0145 (vs. 0.0207 for ODPS) vs. LBLRTM. 
   * Spectral Stability: Completely eliminated spectral "outliers." Zero channels now exceed the 0.1 RMSE threshold (compared to 1,343 channels in the traditional ODPS model) vs LBLRTM. 
   * Performance: Vectorized inference speed is approximately 0.13 ms per profile (8,461 channels), well below the real-time operational requirement.

<img width="2250" height="1200" alt="Image" src="https://github.com/user-attachments/assets/eb3b1175-d1c5-4319-a3a6-075ca794a8b7" />

<img width="3600" height="1800" alt="Image" src="https://github.com/user-attachments/assets/be059902-4ff9-49fd-bdcd-9ad560b0640b" />

<img width="3973" height="3276" alt="Image" src="https://github.com/user-attachments/assets/4fb9042f-9c8c-4a51-a444-c46cb2b283de" />
The Density Correlation plot confirms that the error distribution is highly centered; over 95% of channels fall below the 5:1 Error-to-RMSE ratio. The 'Tail' of outliers (Max Error > 0.3) is confined to a very small population of channels. These typically correspond to regions of high spectral complexity, such as sharp absorption line wings, where the relationship between atmospheric state and transmittance is most non-linear. Even in these 'worst-case' channels, the ML ResNet remains an order of magnitude more stable than the traditional ODPS regression.  


  **Performance:**

Based on our vectorized timing results, the performance is exceptional. Here is the breakdown for processing 1 million IASI profiles (each with 8,461 channels):

  1. Vectorized Inference Performance (Batch Processing)
  Our test showed that onnxruntime can process a batch of 100 profiles in 13.166 ms, which is 0.132 ms per profile.

   * Total Time for 1 Million Profiles: 132,000 ms
   * Final Result: ~132 seconds (2 minutes and 12 seconds)

  2. Individual Profile Performance (Sequential)
  If the profiles are processed one-by-one (e.g., inside a traditional Fortran loop without vectorization), the time per profile was 2.088 ms.

   * Total Time for 1 Million Profiles: 2,088,000 ms
   * Final Result: ~2,088 seconds (34.8 minutes)

  Key Takeaways for the Community:
   * Real-time Operations: Operationally, a single IASI granule contains about 12,000–15,000 profiles. Processing an entire granule would take less than 2 seconds using vectorized inference.
   * Throughput: Even in the "slow" sequential mode, you can process a full million-profile global dataset in just over half an hour on a single CPU core.
   * Hardware Efficiency: These timings were achieved on a standard CPU environment using onnxruntime. If deployed on a GPU, the vectorized batch performance would likely drop to sub-millisecond levels for
     the entire million-profile set, though CPU performance is already more than sufficient for CRTM's needs.

**Comment by BenjaminTJohnson** [comment]
and for ATMS_N20:

<img width="3600" height="1800" alt="Image" src="https://github.com/user-attachments/assets/c1b86825-cf5d-4545-8080-2bf9500935f0" />

<img width="3944" height="3239" alt="Image" src="https://github.com/user-attachments/assets/8c00d02b-7137-4008-b91b-7a53bd220b1f" />

**Comment by BenjaminTJohnson** [comment]
ML Emulator Progress Update: Transition to Physical Predictors (V2)

  To improve the accuracy of the IASI hyperspectral emulator (8,461 channels), we have transitioned from training on raw atmospheric profiles (V1) to training directly on pre-computed physical predictors (V2). This shift leverages the mature physics engine of the traditional CRTM ODPS/OPTRAN regression while retaining the speed of a neural network forward pass.

  1. Key Architectural Shifts
   * Feature Space:
       * V1: Trained on raw atmospheric states ($T$, $P$, $Q_{gas}$). The network had to "learn" the physics of slant-path integration and molecular continuum effects.
       * V2: Trains on the Predictor Matrices ($X$) generated by emulator/odps/predictors.py. These features include pre-integrated path amounts and physically-informed non-linear combinations (e.g., $H_2O / T$, $\sqrt{\text{Path}}$, etc.) used in the standard CRTM.
   * Selective Scaling: Refined the input normalization to use linear scaling for well-mixed gases (like $CO_2$) and logarithmic scaling only for high-dynamic-range absorbers ($H_2O$, $O_3$), aligning the     input distribution with the Beer-Lambert law.
   * Dynamic Channel Weighting: Implemented an automated loss-weighting scheme that amplifies errors in "hard" spectral channels (high-absorption bands) by a factor of up to 50x during training to flatten the
     residual error spectrum.

  2. Performance Comparison (IASI MetOp-A)

    ┌───────────────────────────┬──────────────────────┬─────────────────────────────┬────────────────┐
    │ Metric                    │ ML V1 (Raw Profiles) │ ML V2 (Physical Predictors) │ Improvement    │
    ├───────────────────────────┼──────────────────────┼─────────────────────────────┼────────────────┤
    │ Avg. Channel RMSE         │ 0.0143               │ 0.0091                      │ ~36% reduction │
    │ Max Channel RMSE          │ 0.0367               │ 0.0252                      │ ~31% reduction │
    │ Channels with RMSE < 0.01 │ 13.3%                │ 74.2%                       │ 5.5x increase  │
    └───────────────────────────┴──────────────────────┴─────────────────────────────┴────────────────┘

  3. Qualitative Impact
  The transition to V2 has significantly lowered the error floor across the entire IASI spectral range ($645 - 2765 \text{ cm}^{-1}$). By providing the network with a "linearized" feature space, we successfully removed the large error spikes previously seen in the complex $H_2O$ and $CO_2$ absorption bands.

  4. Integration & Deployment Benefits
  The V2 approach directly aligns the PyTorch model with the CRTM internal data structures. The exported ONNX model now natively ingests the Predictor%ODPS%X arrays already computed within the CRTM_Forward call stack. This removes the need for custom profile-parsing logic in the Fortran bridge, allowing for a cleaner, "drop-in" replacement of the traditional TauCoeff logic.

  Current Status
  The IASI V2 ResNet model (6 blocks, 512 hidden dim) has been fully trained and exported to ONNX. Initial validation against LBLRTM truth shows high spectral fidelity.


<img width="2100" height="1500" alt="Image" src="https://github.com/user-attachments/assets/39ad3b09-2021-499b-b798-7f3e08ec5380" />

1. Spectral Fidelity (Top Panel):
     * ML V1 (Red): Shows significant error spikes, especially in the complex $H_2O$ and $CO_2$ absorption bands.
     * ML V2 (Blue): The error floor is noticeably lower and much flatter across the entire 645-2765 $cm^{-1}$ range. The "noise" in the residuals is significantly reduced by using the pre-integrated physical predictors.
    
2. Error Distribution (Bottom Panel):
     * The "S-curve" shift to the left for V2 is dramatic.
     * While V1 only crossed the 10% threshold at ~0.01 RMSE, V2 has over 74% of its channels below the 0.01 RMSE line.

  Key Takeaway:
  Training on Predictors rather than Profiles provided the neural network with a much more "linearized" and physically structured input space, allowing the ResNet to focus on learning the remaining small non-linear corrections rather than trying to re-learn the physics of slant-path integration from scratch.

**Comment by BenjaminTJohnson** [comment]
<img width="2100" height="1800" alt="Image" src="https://github.com/user-attachments/assets/7e9d6473-c5f0-48f2-8ebe-007724b63e6f" />
